grpc-gcp/src/main/java/com/google/cloud/grpc/multiendpoint/MultiEndpoint.java

```diff
@@ -0,0 +1,201 @@
+    Preconditions.checkNotNull(state);
+    Endpoint endpoint = endpointsMap.get(endpointId);
+    if (endpoint == null) {
+      return;
+    }
+    // If we allow some recovery time.
+    if (EndpointState.UNAVAILABLE.equals(state) && recoveryEnabled) {
+      endpoint.setState(EndpointState.RECOVERING);
+      ScheduledFuture<?> future =
+          executor.schedule(
+              () -> setEndpointStateInternal(endpointId, EndpointState.UNAVAILABLE),
+              recoveryTimeout.toMillis(),
+              MILLISECONDS);
+      endpoint.setChangeStateFuture(future);
+      return;
+    }
+    endpoint.resetStateChangeFuture();
+    endpoint.setState(state);
+    maybeFallback();
+  }
+
+  /**
+   * Provide an updated list of endpoints to MultiEndpoint.
+   *
+   * <p>MultiEndpoint will preserve current endpoints' state and update their priority according to
+   * their new positions.
+   */
+  public synchronized void setEndpoints(List<String> endpoints) {
+    Preconditions.checkNotNull(endpoints);
+    Preconditions.checkArgument(!endpoints.isEmpty(), "Endpoints list must not be empty.");
+
+    // Remove obsolete endpoints.
+    endpointsMap.keySet().retainAll(endpoints);
+
+    // Add new endpoints and update priority.
+    int priority = 0;
+    for (String endpointId : endpoints) {
+      Endpoint existingEndpoint = endpointsMap.get(endpointId);
+      if (existingEndpoint != null) {
+        existingEndpoint.setPriority(priority++);
+        continue;
+      }
+      EndpointState newState =
+          recoveryEnabled ? EndpointState.RECOVERING : EndpointState.UNAVAILABLE;
+      Endpoint newEndpoint = new Endpoint(endpointId, newState, priority++);
+      if (recoveryEnabled) {
+        ScheduledFuture<?> future =
+            executor.schedule(
+                () -> setEndpointStateInternal(endpointId, EndpointState.UNAVAILABLE),
+                recoveryTimeout.toMillis(),
+                MILLISECONDS);
+        newEndpoint.setChangeStateFuture(future);
+      }
+      endpointsMap.put(endpointId, newEndpoint);
+    }
+
+    maybeFallback();
+  }
+
+  private synchronized void maybeFallback() {
+    Optional<Endpoint> topEndpoint =
+        endpointsMap.values().stream()
+            .filter((c) -> c.getState().equals(EndpointState.AVAILABLE))
+            .min(comparingInt(Endpoint::getPriority));
+
+    Endpoint current = endpointsMap.get(currentId);
+    if (current != null && current.getState().equals(EndpointState.RECOVERING)) {
+      // Keep recovering endpoint as current unless a higher priority endpoint became available.
+      if (!topEndpoint.isPresent() || topEndpoint.get().getPriority() >= current.getPriority()) {
+        return;
+      }
+    }
+
+    if (!topEndpoint.isPresent() && current == null) {
+      topEndpoint = endpointsMap.values().stream().min(comparingInt(Endpoint::getPriority));
+    }
+
+    topEndpoint.ifPresent(endpoint -> currentId = endpoint.getId());
```

Comment:
When both top and current endpoints are available, the `topEndpoint` priority is always greater or equal to `current`'s endpoint priority. Then it is always desired to switch to a higher-priority endpoint or in case of equality this is a no-op.